### ISSUE TITLE
Departures from paper wrt discriminator loss, generator first layer and activation functions?

### ISSUE THREAD

**Issue by Elesa** [jsyoon0823/GAIN#3]
Hello,

I have a few comments if I may.

In your code, at line 157 we see that 
D_loss1 = -tf.reduce_mean(M * tf.log(D_prob + 1e-8) + (1-M) * tf.log(1. - D_prob + 1e-8)) * 2

Where does this "*2" come from? (I cannot see it in and/or derive it by the paper)

Also, at line 124 we see that
inputs = tf.concat(axis = 1, values = [inp,m])  # Mask + Data Concatenate

which is also not clarified/stated in the paper as it seems that the generators input is substantially more informative (also input and output dimensions are not the same).

Finally, at the appendix you say that the activation functions at the internal layers that you use are tanh though in your code you use relu; is there a particular reason why you do that?

Many thanks,
Maria Skoularidou

**Comment by jsyoon0823** [comment]
Hello,

Below is my answer for your questions.

1. *2 comes from the fact that in this example, we use 50% missing rate; thus, to compensate that, I put *2. However, note that it is not crucial and you can remove this.

2. As can be seen in Figure 1 in the paper, the inputs of the generators are [x,m,z].
- z comes with x using m*x + (1-m)*z.
- Then, m is concatenated.
- In the case of auto-encoder type of neural networks, you don't need to use the same size of input and output.

3. The activation function for this MNIST example is relu. For the other experiments, we use tanh.

Thanks.

**Comment by Elesa** [comment]
Thanks for your prompt response!

If I may, I still don't see the point of "*2" as, according to the paper, the discriminator is supposed to be trained at the datapoint(s) where b_j=0

Regarding the second point, I do know that generally speaking input and output are not restricted to share the same dimensions (which is also the case for GANs where the dimensionality of input noise is in principal substantially greater than that of the real data)
All I am saying is that based on the paper, I wouldn't suspect that M is needed in such a way (i.e apart from m*x + (1-m)*z)

Best,
Maria

**Comment by jsyoon0823** [comment]
1. You can just ignore "*2". Not much meaningful.
2. Additional "m" is needed in order to distinguish which component comes from x and which component comes from z. Without "m", it is not possible to distinguish these two in terms of the generator.
